core/blockchain/impl/cached_tree.cpp

```diff
@@ -236,19 +178,115 @@ namespace kagome::blockchain {
 
   std::shared_ptr<TreeNode> CachedTree::find(
       const primitives::BlockHash &hash) const {
-    std::queue<std::shared_ptr<TreeNode>> queue;
-    queue.push(root_);
+    if (auto it = nodes_.find(hash); it != nodes_.end()) {
+      return it->second;
+    }
+    return nullptr;
+  }
+
+  std::optional<Reorg> CachedTree::add(
+      const std::shared_ptr<TreeNode> &new_node) {
+    if (nodes_.find(new_node->info.hash) != nodes_.end()) {
+      return std::nullopt;
+    }
+    BOOST_ASSERT(new_node->children.empty());
+    auto parent = wptrMustLock(new_node->weak_parent);
+    auto child_it =
+        std::find(parent->children.begin(), parent->children.end(), new_node);
+    BOOST_ASSERT(child_it == parent->children.end());
```

Comment:
In Release build std::find will work anyway. Let's avoid it 
```suggestion
    BOOST_ASSERT([&](std::find(parent->children.begin(), parent->children.end(), new_node)) == parent->children.end());
```